File: lobster-input-android/app/src/main/java/com/lobster/input/keyboard/pinyin/CustomDictionary.kt
package com.lobster.input.keyboard.pinyin

import android.content.Context

/**
 * 产品内置自定义扩展词库(custom_dict v2,61 万词)。数据访问层,与 [PinyinDictionary] 同构:
 *  - 文件与主词库同格式:`拼音连写key\t词1 lv1 词2 lv2 ...`(key 升序、key 内 lv 降序),
 *    由 tools/keyboard-verify/gen_custom_dict.py 生成,**与主词库零重复**、词频等级同量纲。
 *  - 存储用**字节表**(byte[] + 行偏移索引 + 字节二分,镜像 iOS ByteTable):59 万 key 若用
 *    String 并行数组会撑 ~80MB 热堆,字节表仅 ~25MB(词条文本不解码,命中行才转 String)。
 *  - T9 数字码索引:行号按 key 的数字码字典序排列(IntArray 排列,数字码查询时由 key 字节
 *    即时换算比较,不存储 59 万数字码串)。
 *
 * 加载约定:数据在生成期已保证「全 CJK 2-8 字、拼音可完整切分合法音节、可映射数字码」,
 * 运行时**不做**贪心 allValidSyllables 校验(贪心会误杀 zuoleyinianduo 这类回溯才可切的合法键),
 * 只跳过注释/坏行。由 PinyinEngine 在主词库 READY 后**异步挂载**,挂载前查询按空表返回。
 */
class CustomDictionary {

    @Volatile var isReady: Boolean = false
        private set

    private var bytes: ByteArray = ByteArray(0)
    private var lineStart: IntArray = IntArray(0)
    private var keyLen: IntArray = IntArray(0)
    private var lineEnd: IntArray = IntArray(0)
    /** 行号排列:按 key 的 T9 数字码字典序(同码内按行号即 key 序,稳定)。 */
    private var t9Order: IntArray = IntArray(0)

    var maxDigitLen: Int = 0
        private set

    private val letter2digit = ByteArray(128).also {
        for ((d, letters) in mapOf(
            '2' to "abc", '3' to "def", '4' to "ghi", '5' to "jkl",
            '6' to "mno", '7' to "pqrs", '8' to "tuv", '9' to "wxyz"
        )) for (c in letters) it[c.code] = d.code.toByte()
    }

    fun load(context: Context, assetName: String = "custom_dict.txt") {
        val data = context.assets.open(assetName).use { it.readBytes() }
        val starts = ArrayList<Int>(620_000)
        val klens = ArrayList<Int>(620_000)
        val ends = ArrayList<Int>(620_000)
        var i = 0
        val n = data.size
        while (i < n) {
            val start = i
            var tab = -1
            var j = i
            while (j < n && data[j] != NL) {
                if (data[j] == TAB && tab < 0) tab = j
                j++
            }
            // 跳过注释行与无 TAB 的坏行(数据生成期已保证合法,这里只防御格式损坏)
            if (tab > start && data[start] != HASH) {
                starts.add(start); klens.add(tab - start); ends.add(j)
            }
            i = j + 1
        }
        bytes = data
        lineStart = starts.toIntArray()
        keyLen = klens.toIntArray()
        lineEnd = ends.toIntArray()
        buildT9Order()
        isReady = true
    }

    /** 数字码排列:临时物化数字码串排序(后台线程,峰值 ~30MB 随后 GC),常驻仅 IntArray。 */
    private fun buildT9Order() {
        val count = lineStart.size
        val digitKeys = arrayOfNulls<String>(count)
        var maxLen = 0
        val sb = StringBuilder(24)
        for (idx in 0 until count) {
            sb.setLength(0)
            val s = lineStart[idx]
            val kl = keyLen[idx]
            var valid = true
            for (k in 0 until kl) {
                val b = bytes[s + k].toInt()
                val d = if (b in 0..127) letter2digit[b] else 0
                if (d.toInt() == 0) { valid = false; break }
                sb.append(d.toInt().toChar())
            }
            if (!valid) continue
            digitKeys[idx] = sb.toString()
            if (kl > maxLen) maxLen = kl
        }
        maxDigitLen = maxLen
        val order = (0 until count).filter { digitKeys[it] != null }
            .sortedBy { digitKeys[it]!! }
        t9Order = order.toIntArray()
    }

    // ===== 拼音键查询(字节二分,镜像 iOS ByteTable) =====

    private fun compareKeyAt(index: Int, q: ByteArray): Int {
        val s = lineStart[index]
        val len = keyLen[index]
        val m = minOf(len, q.size)
        for (k in 0 until m) {
            val a = bytes[s + k].toInt() and 0xFF
            val b = q[k].toInt() and 0xFF
            if (a != b) return if (a < b) -1 else 1
        }
        return len.compareTo(q.size)
    }

    private fun lowerBound(q: ByteArray): Int {
        var lo = 0
        var hi = lineStart.size
        while (lo < hi) {
            val mid = (lo + hi) ushr 1
            if (compareKeyAt(mid, q) < 0) lo = mid + 1 else hi = mid
        }
        return lo
    }

    private fun keyStartsWith(index: Int, q: ByteArray): Boolean {
        if (keyLen[index] < q.size) return false
        val s = lineStart[index]
        for (k in q.indices) if (bytes[s + k] != q[k]) return false
        return true
    }

    private fun keyAt(index: Int): String =
        String(bytes, lineStart[index], keyLen[index], Charsets.UTF_8)

    /** 解析行内候选:`词1 lv1 词2 lv2 ...`(与主词库 SortedTable.wordsAt 同语义)。 */
    private fun wordsAt(index: Int, limit: Int = Int.MAX_VALUE): List<Pair<String, Int>> {
        val out = ArrayList<Pair<String, Int>>()
        var p = lineStart[index] + keyLen[index] + 1
        val end = lineEnd[index]
        while (p < end && out.size < limit) {
            var q = p
            while (q < end && bytes[q] != SP) q++
            if (q >= end) break
            val word = String(bytes, p, q - p, Charsets.UTF_8)
            var r = q + 1
            var lv = 0
            var any = false
            while (r < end && bytes[r] != SP) {
                val c = bytes[r].toInt() - '0'.code
                if (c in 0..9) { lv = lv * 10 + c; any = true }
                r++
            }
            if (word.isNotEmpty() && any) out.add(word to lv)
            p = r + 1
        }
        return out
    }

    /** 整键精确查词(与主词库 exactWords 同语义)。 */
    fun exactWords(key: String): List<Pair<String, Int>> {
        if (!isReady || key.isEmpty()) return emptyList()
        val q = key.toByteArray(Charsets.UTF_8)
        val i = lowerBound(q)
        if (i >= lineStart.size || compareKeyAt(i, q) != 0) return emptyList()
        return wordsAt(i)
    }

    fun bestWord(key: String): Pair<String, Int>? {
        if (!isReady || key.isEmpty()) return null
        val q = key.toByteArray(Charsets.UTF_8)
        val i = lowerBound(q)
        if (i >= lineStart.size || compareKeyAt(i, q) != 0) return null
        return wordsAt(i, 1).firstOrNull()
    }

    /** 前缀补全:每个更长同前缀 key 取首选词,返回 (词, 等级, key长)(与主词库 prefixWords 同语义)。 */
    fun prefixWords(prefix: String, limit: Int): List<Triple<String, Int, Int>> {
        if (!isReady || prefix.isEmpty()) return emptyList()
        val q = prefix.toByteArray(Charsets.UTF_8)
        val out = ArrayList<Triple<String, Int, Int>>()
        var i = lowerBound(q)
        while (i < lineStart.size && out.size < limit) {
            if (!keyStartsWith(i, q)) break
            if (keyLen[i] != q.size) {
                wordsAt(i, 1).firstOrNull()?.let { (w, lv) -> out.add(Triple(w, lv, keyLen[i])) }
            }
            i++
        }
        return out
    }

    // ===== T9 数字码查询(排列二分,数字码由 key 字节即时换算) =====

    /** key(行 index)的数字码 与 查询数字码 逐位比较;prefixOnly 时 key 更长视为相等(前缀命中)。 */
    private fun compareDigitsAt(index: Int, digits: String, prefixOnly: Boolean): Int {
        val s = lineStart[index]
        val len = keyLen[index]
        val m = minOf(len, digits.length)
        for (k in 0 until m) {
            val b = bytes[s + k].toInt()
            val d = (if (b in 0..127) letter2digit[b] else 0).toInt().toChar()
            val q = digits[k]
            if (d != q) return if (d < q) -1 else 1
        }
        if (prefixOnly && len >= digits.length) return 0
        return len.compareTo(digits.length)
    }

    private fun t9LowerBound(digits: String): Int {
        var lo = 0
        var hi = t9Order.size
        while (lo < hi) {
            val mid = (lo + hi) ushr 1
            if (compareDigitsAt(t9Order[mid], digits, prefixOnly = false) < 0) lo = mid + 1 else hi = mid
        }
        return lo
    }

    /** 数字码精确查词:跨同码拼音合并,返回 (词, 等级, 拼音key)(与主词库 t9ExactWords 同语义)。 */
    fun t9ExactWords(digits: String): List<Triple<String, Int, String>> {
        if (!isReady || digits.isEmpty()) return emptyList()
        val out = ArrayList<Triple<String, Int, String>>()
        var i = t9LowerBound(digits)
        while (i < t9Order.size) {
            val ref = t9Order[i]
            if (compareDigitsAt(ref, digits, prefixOnly = false) != 0) break
            val key = keyAt(ref)
            for ((w, lv) in wordsAt(ref)) out.add(Triple(w, lv, key))
            i++
        }
        return out
    }

    /** 数字码前缀补全:每个更长同前缀码取首选词(与主词库 t9PrefixWords 同语义)。 */
    fun t9PrefixWords(digits: String, limit: Int): List<PinyinDictionary.T9PrefixHit> {
        if (!isReady || digits.isEmpty()) return emptyList()
        val out = ArrayList<PinyinDictionary.T9PrefixHit>()
        var i = t9LowerBound(digits)
        while (i < t9Order.size && out.size < limit) {
            val ref = t9Order[i]
            if (compareDigitsAt(ref, digits, prefixOnly = true) != 0) break
            if (keyLen[ref] != digits.length) {
                wordsAt(ref, 1).firstOrNull()?.let { (w, lv) ->
                    out.add(PinyinDictionary.T9PrefixHit(w, lv, keyLen[ref], keyAt(ref)))
                }
            }
            i++
        }
        return out
    }

    private companion object {
        private const val NL = '\n'.code.toByte()
        private const val TAB = '\t'.code.toByte()
        private const val SP = ' '.code.toByte()
        private const val HASH = '#'.code.toByte()
    }
}

File: lobster-input-ios/LobsterKeyboard/Pinyin/CustomDictionary.swift
import Foundation

/// 产品内置自定义扩展词库(custom_dict v2,61 万词)。数据访问层,与 PinyinDictionary 同构:
///  - 文件与主词库同格式:`拼音连写key\t词1 lv1 词2 lv2 ...`(key 升序、key 内 lv 降序),
///    由 tools/keyboard-verify/gen_custom_dict.py 生成,**与主词库零重复**、词频等级同量纲。
///  - 存储照抄主词库 ByteTable 模式:Data(.mappedIfSafe) 内存映射 + 紧凑行偏移索引 + 字节二分
///    (59 万 key 若解码成 String 并行数组会撑爆键盘扩展内存上限,字节表词条文本不解码、
///    命中行才转 String;mmap 为只读文件页,不计 dirty 内存)。
///  - T9 数字码索引:行号按 key 的数字码字典序排列([Int32] 排列,数字码查询/排序时由 key
///    字节即时换算比较,**不物化 59 万数字码串**——物化峰值 ~40MB 会顶到扩展 jetsam 线)。
///
/// 加载约定:数据在生成期已保证「全 CJK 2-8 字、拼音可完整切分合法音节、可映射数字码」,
/// 运行时**不做**贪心 allValidSyllables 校验(贪心会误杀 zuoleyinianduo 这类回溯才可切的合法键),
/// 只跳过注释/坏行。由 PinyinEngine 在主词库 READY 后**异步挂载**,挂载前查询按空表返回。
final class CustomDictionary {

    private(set) var isReady = false

    private var bytes: Data = Data()
    private var lineStart: [UInt32] = []
    private var keyLen: [UInt16] = []
    private var lineEnd: [UInt32] = []
    /// 行号排列:按 key 的 T9 数字码字典序(同码内按行号即 key 升序,稳定)。
    private var t9Order: [Int32] = []

    /// 自定义词最长数字码(T9 词层循环上限,biang=14 位不受 maxT9WordDigits 截断)。
    private(set) var maxDigitLen = 0

    private let letter2digit: [UInt8] = {
        var m = [UInt8](repeating: 0, count: 128)
        for c in "abc".utf8 { m[Int(c)] = UInt8(ascii: "2") }
        for c in "def".utf8 { m[Int(c)] = UInt8(ascii: "3") }
        for c in "ghi".utf8 { m[Int(c)] = UInt8(ascii: "4") }
        for c in "jkl".utf8 { m[Int(c)] = UInt8(ascii: "5") }
        for c in "mno".utf8 { m[Int(c)] = UInt8(ascii: "6") }
        for c in "pqrs".utf8 { m[Int(c)] = UInt8(ascii: "7") }
        for c in "tuv".utf8 { m[Int(c)] = UInt8(ascii: "8") }
        for c in "wxyz".utf8 { m[Int(c)] = UInt8(ascii: "9") }
        return m
    }()

    func load() {
        guard let url = Bundle(for: CustomDictionary.self).url(forResource: "custom_dict", withExtension: "txt"),
              let data = try? Data(contentsOf: url, options: .mappedIfSafe) else { return }
        var starts: [UInt32] = []
        var klens: [UInt16] = []
        var ends: [UInt32] = []
        starts.reserveCapacity(620_000); klens.reserveCapacity(620_000); ends.reserveCapacity(620_000)
        let n = data.count
        // 61 万行/16.7MB:扫描走裸指针(Data 逐字节下标在扩展里慢一个量级,拖冷挂载时长)
        data.withUnsafeBytes { (buf: UnsafeRawBufferPointer) in
            let p = buf.bindMemory(to: UInt8.self)
            var i = 0
            while i < n {
                let start = i
                var tab = -1
                var j = i
                while j < n && p[j] != 0x0A {
                    if p[j] == 0x09 && tab < 0 { tab = j }
                    j += 1
                }
                // 跳过注释行与无 TAB 的坏行(数据生成期已保证合法,这里只防御格式损坏)
                if tab > start && p[start] != 0x23 {
                    starts.append(UInt32(start)); klens.append(UInt16(min(tab - start, 65535))); ends.append(UInt32(j))
                }
                i = j + 1
            }
        }
        bytes = data
        lineStart = starts
        keyLen = klens
        lineEnd = ends
        buildT9Order()
        isReady = true
    }

    /// 数字码排列:比较器由 key 字节即时换算数字码(后台线程一次性 ~11M 次比较,
    /// 换取零物化内存——键盘扩展内存上限严苛,临时数字码串数组的分配峰值不可接受)。
    private func buildT9Order() {
        let count = lineStart.count
        var order: [Int32] = []
        order.reserveCapacity(count)
        var maxLen = 0
        bytes.withUnsafeBytes { (buf: UnsafeRawBufferPointer) in
            let p = buf.bindMemory(to: UInt8.self)
            for idx in 0 ..< count {
                let s = Int(lineStart[idx]); let kl = Int(keyLen[idx])
                var valid = true
                for k in 0 ..< kl {
                    let b = p[s + k]
                    if b >= 128 || letter2digit[Int(b)] == 0 { valid = false; break }
                }
                if !valid { continue }
                if kl > maxLen { maxLen = kl }
                order.append(Int32(idx))
            }
            order.sort { a, b in
                let c = compareDigitRows(p, Int(a), Int(b))
                return c != 0 ? c < 0 : a < b
            }
        }
        maxDigitLen = maxLen
        t9Order = order
    }

    /// 两行 key 的数字码字典序比较(排序用;key 已验证全部可映射数字)。
    private func compareDigitRows(_ p: UnsafeBufferPointer<UInt8>, _ ia: Int, _ ib: Int) -> Int {
        let sa = Int(lineStart[ia]); let la = Int(keyLen[ia])
        let sb = Int(lineStart[ib]); let lb = Int(keyLen[ib])
        let m = min(la, lb)
        var k = 0
        while k < m {
            let da = letter2digit[Int(p[sa + k])]
            let db = letter2digit[Int(p[sb + k])]
            if da != db { return da < db ? -1 : 1 }
            k += 1
        }
        if la == lb { return 0 }
        return la < lb ? -1 : 1
    }

    // MARK: 拼音键查询(字节二分,照抄 PinyinDictionary.ByteTable 模式)

    private func compareKey(at index: Int, _ q: [UInt8]) -> Int {
        let s = Int(lineStart[index]); let len = Int(keyLen[index])
        let m = min(len, q.count)
        var k = 0
        while k < m {
            let a = bytes[s + k], b = q[k]
            if a != b { return a < b ? -1 : 1 }
            k += 1
        }
        if len == q.count { return 0 }
        return len < q.count ? -1 : 1
    }

    private func lowerBound(_ q: [UInt8]) -> Int {
        var lo = 0, hi = lineStart.count
        while lo < hi {
            let mid = (lo + hi) >> 1
            if compareKey(at: mid, q) < 0 { lo = mid + 1 } else { hi = mid }
        }
        return lo
    }

    private func keyStartsWith(_ index: Int, _ q: [UInt8]) -> Bool {
        if Int(keyLen[index]) < q.count { return false }
        let s = Int(lineStart[index])
        for k in 0 ..< q.count where bytes[s + k] != q[k] { return false }
        return true
    }

    private func keyAt(_ index: Int) -> String {
        let s = Int(lineStart[index])
        return String(decoding: bytes[s ..< s + Int(keyLen[index])], as: UTF8.self)
    }

    /// 解析行内候选:`词1 lv1 词2 lv2 ...`(与主词库 wordsAt 同语义;limit 供取首选词短路)。
    private func wordsAt(_ index: Int, limit: Int = Int.max) -> [(String, Int)] {
        let s = Int(lineStart[index]) + Int(keyLen[index]) + 1
        let e = Int(lineEnd[index])
        guard s < e else { return [] }
        let parts = String(decoding: bytes[s ..< e], as: UTF8.self).split(separator: " ")
        var out: [(String, Int)] = []
        var i = 0
        while i + 1 < parts.count && out.count < limit {
            let w = String(parts[i]); let lv = Int(parts[i + 1]) ?? 0
            if !w.isEmpty { out.append((w, lv)) }
            i += 2
        }
        return out
    }

    /// 整键精确查词(与主词库 exactWords 同语义)。
    func exactWords(_ key: String) -> [(String, Int)] {
        guard isReady, !key.isEmpty else { return [] }
        let q = [UInt8](key.utf8)
        let i = lowerBound(q)
        if i >= lineStart.count || compareKey(at: i, q) != 0 { return [] }
        return wordsAt(i)
    }

    func bestWord(_ key: String) -> (String, Int)? {
        guard isReady, !key.isEmpty else { return nil }
        let q = [UInt8](key.utf8)
        let i = lowerBound(q)
        if i >= lineStart.count || compareKey(at: i, q) != 0 { return nil }
        return wordsAt(i, limit: 1).first
    }

    /// 前缀补全:每个更长同前缀 key 取首选词,返回 (词, 等级, key长)(与主词库 prefixWords 同语义)。
    func prefixWords(_ prefix: String, _ limit: Int) -> [(String, Int, Int)] {
        guard isReady, !prefix.isEmpty else { return [] }
        let q = [UInt8](prefix.utf8)
        var out: [(String, Int, Int)] = []
        var i = lowerBound(q)
        while i < lineStart.count && out.count < limit {
            if !keyStartsWith(i, q) { break }
            if Int(keyLen[i]) != q.count, let first = wordsAt(i, limit: 1).first {
                out.append((first.0, first.1, Int(keyLen[i])))
            }
            i += 1
        }
        return out
    }

    // MARK: T9 数字码查询(排列二分,数字码由 key 字节即时换算)

    /// key(行 index)的数字码与查询数字码逐位比较;prefixOnly 时 key 更长视为相等(前缀命中)。
    private func compareDigits(at index: Int, _ q: [UInt8], prefixOnly: Bool) -> Int {
        let s = Int(lineStart[index]); let len = Int(keyLen[index])
        let m = min(len, q.count)
        var k = 0
        while k < m {
            let b = bytes[s + k]
            let d = b < 128 ? letter2digit[Int(b)] : 0
            if d != q[k] { return d < q[k] ? -1 : 1 }
            k += 1
        }
        if prefixOnly && len >= q.count { return 0 }
        if len == q.count { return 0 }
        return len < q.count ? -1 : 1
    }

    private func t9LowerBound(_ q: [UInt8]) -> Int {
        var lo = 0, hi = t9Order.count
        while lo < hi {
            let mid = (lo + hi) >> 1
            if compareDigits(at: Int(t9Order[mid]), q, prefixOnly: false) < 0 { lo = mid + 1 } else { hi = mid }
        }
        return lo
    }

    /// 数字码精确查词:跨同码拼音合并,返回 (词, 等级, 拼音key)(与主词库 t9ExactWords 同语义)。
    func t9ExactWords(_ digits: String) -> [(String, Int, String)] {
        guard isReady, !digits.isEmpty else { return [] }
        let q = [UInt8](digits.utf8)
        var out: [(String, Int, String)] = []
        var i = t9LowerBound(q)
        while i < t9Order.count {
            let ref = Int(t9Order[i])
            if compareDigits(at: ref, q, prefixOnly: false) != 0 { break }
            let key = keyAt(ref)
            for (w, lv) in wordsAt(ref) { out.append((w, lv, key)) }
            i += 1
        }
        return out
    }

    /// 数字码前缀补全:每个更长同前缀码取首选词(与主词库 t9PrefixWords 同语义)。
    func t9PrefixWords(_ digits: String, _ limit: Int) -> [PinyinDictionary.T9PrefixHit] {
        guard isReady, !digits.isEmpty else { return [] }
        let q = [UInt8](digits.utf8)
        var out: [PinyinDictionary.T9PrefixHit] = []
        var i = t9LowerBound(q)
        while i < t9Order.count && out.count < limit {
            let ref = Int(t9Order[i])
            if compareDigits(at: ref, q, prefixOnly: true) != 0 { break }
            if Int(keyLen[ref]) != q.count, let first = wordsAt(ref, limit: 1).first {
                out.append(PinyinDictionary.T9PrefixHit(word: first.0, level: first.1, keyDigitLen: Int(keyLen[ref]), pinyinKey: keyAt(ref)))
            }
            i += 1
        }
        return out
    }
}
